rust/hyperlane-base/src/contract_sync/cursor.rs

```diff
@@ -167,99 +241,115 @@ impl ContractSyncCursor<HyperlaneMessage> for ForwardMessageSyncCursor {
+        mode: IndexMode,
+        next_sequence: u32,
+        synced: bool,
+    ) -> Self {
+        Self {
+            cursor: MessageSyncCursor::new(
+                indexer,
+                db,
+                SyncState::new(
+                    chunk_size,
+                    start_block,
+                    next_block,
+                    mode,
+                    next_sequence,
+                    SyncDirection::Backward,
+                ),
+            ),
+            synced,
+        }
+    }
+
+    async fn get_next_range(&mut self) -> ChainResult<Option<RangeInclusive<u32>>> {
         // Check if any new messages have been inserted into the DB,
         // and update the cursor accordingly.
         while !self.synced {
             if self
                 .cursor
-                .retrieve_message_by_nonce(self.cursor.next_nonce)
+                .retrieve_message_by_nonce(self.cursor.sync_state.next_sequence)
                 .await
                 .is_none()
             {
                 break;
             };
-            // If we found nonce zero or hit block zero, we are done rewinding.
-            if self.cursor.next_nonce == 0 || self.cursor.next_block == 0 {
+            // If we found sequence zero or hit block zero, we are done rewinding.
+            if self.cursor.sync_state.next_sequence == 0 || self.cursor.sync_state.next_block == 0 {
                 self.synced = true;
                 break;
             }
 
             if let Some(block_number) = self
                 .cursor
-                .retrieve_dispatched_block_number(self.cursor.next_nonce)
+                .retrieve_dispatched_block_number(self.cursor.sync_state.next_sequence)
                 .await
             {
                 // It's possible that eth_getLogs dropped logs from this block, therefore we cannot do block_number - 1.
-                self.cursor.next_block = block_number;
+                self.cursor.sync_state.next_block = block_number;
             }
 
-            self.cursor.next_nonce = self.cursor.next_nonce.saturating_sub(1);
+            self.cursor.sync_state.next_sequence =
+                self.cursor.sync_state.next_sequence.saturating_sub(1);
         }
         if self.synced {
-            return None;
+            return Ok(None);
         }
 
         // Just keep going backwards.
-        let to = self.cursor.next_block;
-        let from = to.saturating_sub(self.cursor.chunk_size);
-        self.cursor.next_block = from.saturating_sub(1);
-
-        let next_nonce = self.cursor.next_nonce;
-
-        let range = match self.mode {
-            IndexMode::Block => BlockRange(from..=to),
-            IndexMode::Sequence => {
-                SequenceRange(next_nonce.saturating_sub(MAX_SEQUENCE_RANGE)..=next_nonce)
-            }
-        };
-
-        Some(range)
+        let (count, tip) = self.cursor.indexer.sequence_and_tip().await?;
+        self.cursor.sync_state.get_next_range(count, tip).await
```

Comment:
Does this return a correct range when sequence indexing? Looking at SyncState's `get_next_range` I don't see anything when the sync direction is backward that results in a backward sequence range being returned.

Wondering if things functionally still work because of the `self.cursor.sync_state.next_sequence = self.cursor.sync_state.next_sequence.saturating_sub(1);` up above, but instead of returning a range like [95, 100], then [90, 95], etc, what's returned is [99, 104], then [98, 103], etc (so the start range does go down, but messages end up being indexed multiple times)

Possible i'm misreading things though